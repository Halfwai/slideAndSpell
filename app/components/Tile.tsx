import { View, Text, StyleSheet, TouchableOpacity, Animated, PanResponder } from 'react-native';
import { Direction } from '../constants/enums';
import { useEffect, useRef, useState } from 'react';

export default function Tile(props : {value : number, position : {y : number, x: number}, slidable : Direction, switch : Function}) {
    // console.log(props);

    const tilePosition = useRef(new Animated.ValueXY({x: props.position.x * 75, y: props.position.y * 75})).current;
    const slidableRef = useRef(props.slidable);
    const [colour, setColour] = useState('#AAA');

    useEffect(() => {
        slidableRef.current = props.slidable;
    }, [props.slidable]);


    const panResponder = useRef(
        PanResponder.create({
            onMoveShouldSetPanResponder: () => true,
            onPanResponderMove: (e, gestureState) => {
                setColour('#CBE6F7');
                if (slidableRef.current === Direction.RIGHT) {
                    const gestureX = Math.max(Math.min(gestureState.dx, 75), 0);
                    tilePosition.x.setValue(props.position.x * 75 + gestureX);
                } else if (slidableRef.current === Direction.LEFT) {
                    const gestureX = Math.min(Math.max(gestureState.dx, -75), 0);
                    tilePosition.x.setValue(props.position.x * 75 + gestureX);
                } else if (slidableRef.current === Direction.UP) {
                    const gestureY = Math.min(Math.max(gestureState.dy, -75), 0);
                    tilePosition.y.setValue(props.position.y * 75 + gestureY);
                } else if (slidableRef.current === Direction.DOWN) {
                    const gestureY = Math.max(Math.min(gestureState.dy, 75), 0);
                    tilePosition.y.setValue(props.position.y * 75 + gestureY);
                }
            },
            onPanResponderRelease: (e, gestureState) => {
                setColour('#AAA');
                if (slidableRef.current === Direction.RIGHT) {
                    if (gestureState.dx <= 0) return;
                    moveRight();
                } else if (slidableRef.current === Direction.LEFT) {
                    if (gestureState.dx >= 0) return;
                    moveLeft();
                } else if (slidableRef.current === Direction.UP) {
                    if (gestureState.dy >= 0) return;
                    moveUp();
                } else if (slidableRef.current === Direction.DOWN) {
                    if (gestureState.dy <= 0) return;
                    moveDown();
                }
            },
        })
    ).current;   


    useEffect(() => {
        tilePosition.setValue({x: props.position.x * 75, y: props.position.y * 75});
    }, [props.position]);

    function moveRight() {
        Animated.spring(tilePosition.x, {
            toValue: props.position.x * 75 + 75,
            bounciness: 0,
            speed: 50,
            useNativeDriver: true
        }).start(({finished}) => {
            if(!finished) return;
            props.switch(props.position.y, props.position.x);
        });
    }
    function moveLeft() {
        Animated.spring(tilePosition.x, {
            toValue: props.position.x * 75 - 75,
            bounciness: 0,
            speed: 50,
            useNativeDriver: true
        }).start(({finished}) => {
            if(!finished) return;
            props.switch(props.position.y, props.position.x);
        });
    }
    function moveUp() {
        Animated.spring(tilePosition.y, {
            toValue: props.position.y * 75 - 75,
            bounciness: 0,
            speed: 50,
            useNativeDriver: true
        }).start(({finished}) => {
            if(!finished) return;
            props.switch(props.position.y, props.position.x);
        });
    }
    function moveDown() {
        Animated.spring(tilePosition.y, {
            toValue: props.position.y * 75 + 75,
            bounciness: 0,
            speed: 50,
            useNativeDriver: true
        }).start(({finished}) => {
            if(!finished) return;
            props.switch(props.position.y, props.position.x);
        });
    }
    return (
        <>
            { props.value === 0 ? null : 
                <Animated.View 
                    style={[
                        styles.tile,
                        {
                            transform: [{ translateX: tilePosition.x }, { translateY: tilePosition.y }],
                            backgroundColor: colour,
                        }
                    ]}
                    {...panResponder.panHandlers}
                >
                    <Text style={styles.text}>{props.value}</Text>
                </Animated.View>
            }

        </>
    );
}

const styles = StyleSheet.create({
    tile: {
        width: 75,
        height: 75,
        position: 'absolute',
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 1,
        borderRadius: 10,
    },
    text: {
        textAlign: 'center',
        fontSize: 24,
    }
});