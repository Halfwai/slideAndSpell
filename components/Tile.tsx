// Import necessary libraries
import { Text, StyleSheet, Animated, PanResponder, ImageSize } from 'react-native';
import { useEffect, useRef } from 'react';
import React from 'react';
import * as Haptics from 'expo-haptics';

// Import the Direction enum
import { Direction } from '@/constants/enums';

// Define the Tile component
export default function Tile(props: { value: string | number, position: { y: number, x: number }, spaceSize : number, tileSize : number, slidable: Direction, switch: Function, resetBoard: Function, valid: boolean }) {
    // Define the tilePosition, slidableRef, and colour variables
    const tilePosition = useRef(new Animated.ValueXY({ x: props.position.x * (props.spaceSize), y: props.position.y * (props.spaceSize) })).current;
    const slidableRef = useRef(props.slidable);
    const colour = useRef(new Animated.Value(0)).current;

    useEffect(() => {
        let newColour = props.valid ? 3 : 0;
        changeColour(newColour);

    }, [props.valid]);

    // Update the slidableRef when props.slidable changes
    useEffect(() => {
        slidableRef.current = props.slidable;
    }, [props.slidable]);

    // Define the panResponder
    const panResponder = useRef(
        PanResponder.create({       
            onMoveShouldSetPanResponder: () => true,
            // Haptic feedback when the user touches the tile
            onPanResponderGrant: () => {
                (async () => {
                    await Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
                })();
            },
            onPanResponderMove: (e, gestureState) => {
                // Change the colour of the tile when the user moves it. If the tile can be moved it will be blue, otherwise it will be red.
                if (slidableRef.current === Direction.FALSE) {
                    changeColour(2);
                } else {
                    changeColour(1);
                }
                
                // Move the tile in the direction the user is moving it. Tile position is locked to the empty space.
                if (slidableRef.current === Direction.RIGHT) {
                    const gestureX = Math.max(Math.min(gestureState.dx, props.spaceSize), 0);
                    tilePosition.x.setValue(props.position.x * (props.spaceSize) + gestureX);
                } else if (slidableRef.current === Direction.LEFT) {
                    const gestureX = Math.min(Math.max(gestureState.dx, - props.spaceSize), 0);
                    tilePosition.x.setValue(props.position.x * (props.spaceSize) + gestureX);
                } else if (slidableRef.current === Direction.UP) {
                    const gestureY = Math.max(Math.min(gestureState.dy, 0), - props.spaceSize);
                    tilePosition.y.setValue(props.position.y * (props.spaceSize) + gestureY);
                } else if (slidableRef.current === Direction.DOWN) {
                    const gestureY = Math.min(Math.max(gestureState.dy, 0), props.spaceSize);
                    tilePosition.y.setValue(props.position.y * (props.spaceSize) + gestureY);;
                }
            },
            // Move the tile to the empty space when the user releases it
            onPanResponderRelease: (e, gestureState) => {
                if (slidableRef.current === Direction.RIGHT) {
                    if (gestureState.dx <= 0) return;
                    moveTile({ x: props.position.x *  props.spaceSize +  props.spaceSize, y: props.position.y *  props.spaceSize });
                } else if (slidableRef.current === Direction.LEFT) {
                    if (gestureState.dx >= 0) return;
                    moveTile({ x: props.position.x *  props.spaceSize -  props.spaceSize, y: props.position.y *  props.spaceSize });
                } else if (slidableRef.current === Direction.UP) {
                    if (gestureState.dy >= 0) return;
                    moveTile({ x: props.position.x *  props.spaceSize, y: props.position.y *  props.spaceSize -  props.spaceSize });
                } else if (slidableRef.current === Direction.DOWN) {
                    if (gestureState.dy <= 0) return;
                    moveTile({ x: props.position.x *  props.spaceSize, y: props.position.y *  props.spaceSize +  props.spaceSize });
                }
            },
        })
    ).current;

    // interpolateColour is used to change the colour of the tile based on the value of the animated colour variable
    const interpolateColour = colour.interpolate({
        inputRange: [0, 1, 2, 3],
        outputRange: ['#AAA', '#CBE6F7', "#CB4C4E", "#8bf0bc"]
    });

    // changeColour is used to change the colour of the tile
    const changeColour = (target: number) => {
        Animated.timing(colour, {
            toValue: target,
            duration: 1,
            useNativeDriver: true
        }).start();
    }

    // useEffect to update the tilePosition when props.position changes
    useEffect(() => {
        tilePosition.setValue({ x: props.position.x * (props.spaceSize), y: props.position.y * (props.spaceSize) });
    }, [props.position]);

    
    // moveTile function moves the tile to the destination
    function moveTile(destination: { x: number, y: number }) {
        Animated.spring(tilePosition, {
            toValue: {x: destination.x, y: destination.y},
            // bounciness set to 0 to prevent the tile from bouncing into other tiles
            bounciness: 0,
            // speed set to 1000 to make the animation feel more responsive
            speed: 1000,
            useNativeDriver: true
        }).start(({ finished }) => {
            if (!finished) return;
            // switch the tile with the empty space in the board when the animation is finished
            const newBoard = props.switch(props.position.y, props.position.x);
            if (newBoard === undefined) return;
            props.resetBoard(newBoard);
        });
    }

    return (
        <>
            {props.value === "0" ? null :
                <Animated.View
                    style={[
                        styles(props.tileSize).tile,
                        {
                            transform: [{ translateX: tilePosition.x }, { translateY: tilePosition.y }],
                            backgroundColor: interpolateColour,
                        }
                    ]}
                    {...panResponder.panHandlers}
                >
                    <Text style={styles().text}>{props.value}</Text>
                </Animated.View>
            }
        </>
    );
}

// Define the styles for the Tile component
const styles = (props : number | undefined) => StyleSheet.create({
    tile: {
        width: props,
        height: props,
        margin: 1,
        position: 'absolute',
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 1,
        borderRadius: 10,
    },
    text: {
        textAlign: 'center',
        fontSize: 24,
    }
});