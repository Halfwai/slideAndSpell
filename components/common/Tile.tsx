import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { RFPercentage } from "react-native-responsive-fontsize";

// Setup props
interface TileProps {
    letter: string,
}

export default function Tile(props: TileProps) {
    return (
        <View style={styles.container}>
            <Text style={styles.text}>{props.letter}</Text>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 1,
        borderRadius: 10,
        width: "100%",
        height: "100%",
    },
    text: {
        textAlign: 'center',
        fontSize: RFPercentage(3.8),
        fontFamily: 'Postino_std',
    }
});