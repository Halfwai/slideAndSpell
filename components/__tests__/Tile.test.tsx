import { render } from '@testing-library/react-native';
import Tile from '@/components/common/Tile';

it(`Tile Component renders correctly`, () => {
    const { toJSON } = render(
        <Tile
            letter='A'
        />
    );    
    expect(toJSON()).toMatchSnapshot();
});