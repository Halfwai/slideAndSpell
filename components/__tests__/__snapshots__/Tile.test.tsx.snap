// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile Component renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "borderRadius": 10,
      "borderWidth": 1,
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "fontFamily": "Postino_std",
        "fontSize": 24,
        "textAlign": "center",
      }
    }
  >
    A
  </Text>
</View>
`;
